### ISSUE TITLE
Internal error after forced shutdown

### ISSUE THREAD

**Issue by peacememories** [martinvonz/jj#3370]
## Description

I just had to reset my machine because of RAM starvation (not related to Jujutsu) after I moved through the history with `jj edit`. The command had already completed, but apparently some parts of the change were not yet written to persistent storage, because after rebooting all `jj` commands spit out an error.

## Steps to Reproduce the Problem

1. run `jj edit <change>`
1. wait for the command to complete
1. power off the machine

## Expected Behavior

Some data might be lost, but jujutsu is still in a workable state

## Actual Behavior

Every `jj` command results in the output

```
Internal error: Failed to load the repo: Error when reading object  of type view: Is a directory (os error 21)
```

## Specifications

- Platform: nixOS 23.11
- Version: jj 0.13.0

**Comment by peacememories** [comment]
I just updated to jj 0.15.1 and the problem persists, though I cannot say if jj 0.15.1 would have destroyed the repo in the same way.

I moved my `.jj` directory to a backup location and initialized a new one, but the commits after my current one just seem to be gone. I can't even find them in gits reflog, because that only contains `checkout: moving` entries. I'm guessing this is because jujutsu creates commits in a way that do not touch gits reflog.

**Comment by yuja** [comment]
> ```
> Internal error: Failed to load the repo: Error when reading object  of type view: Is a directory (os error 21)
> ```

Appears that the operation object (referred by `.jj/repo/op_heads/heads`) corrupted.
Maybe you can find recent non-corrupted operation file in `.jj/repo/op_store/operations`, and create an empty `.jj/repo/op_heads/heads/{operation_file_name}` file to recover.

> but the commits after my current one just seem to be gone. I can't even find them in gits reflog,

reflog wouldn't help because `refs/jj/keep` isn't tracked. `git checkout refs/jj/keep/{hash}` might help to recover head commit one by one.

**Comment by peacememories** [comment]
Is there a way to show more information about the `operations` files? I'm not really sure how to navigate the files (which are just named after hashes as far as I can tell) to find a non-corrupted version of the change Im looking for.

Can I somehow list what's in `refs/jj/keep`? I unfortunately don't know the hash of the latest change.

I didn't lose *much* work, so I can just do it again, but I'm obviously interested in learning how to fix this for the future :)

**Comment by yuja** [comment]
> Is there a way to show more information about the `operations` files?

There's `jj debug operation`, but it might crash if the content is corrupted.
If you have `protoc`, `protoc --decode op_store.Operation $jj_source/lib/src/protos/op_store.proto < $operation_file` also works.

> I'm not really sure how to navigate the files (which are just named after hashes as far as I can tell) to find a non-corrupted version of the change Im looking for.

I would try a few recent files.

> Can I somehow list what's in `refs/jj/keep`?

Try `git show-ref`

**Comment by peacememories** [comment]
Ah, I tried the second most recent file and now have a working repo again. The newest commit still seems to have disappeared, but at least I have all my other changes back^^

**Comment by yuja** [comment]
Closed as dupe of #1907. Perhaps, fsync() stuff will help improve data integrity.
